inotify-tools: add fanotify support and refactor package expression

rogarb opened the pull request:
This allows to build additional utils, fsnotifywait and fsnotifywatch, which use the fanotify kernel API.

The fanotify support is enabled by default since the default NixOS kernel supports the fanotify API.

The derivation is now built using the `finalAttrs` syntax (see [documentation](https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes)) and the fetcher uses the `hash` attribute (see [documentation](https://nixos.org/manual/nixpkgs/unstable/#fetchfromgithub)).

###### Description of changes

<!--
For package updates please link to a changelog or describe changes, this helps your fellow maintainers discover breaking updates.
For new packages please briefly describe the package or provide a link to its homepage.
-->

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- [ ] For non-Linux: Is `sandbox = true` set in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://nixos.org/manual/nixpkgs/unstable/#sec-package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://nixos.org/manual/nixpkgs/unstable/#ssec-nixos-tests-linking) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [23.05 Release Notes (or backporting 22.11 Release notes)](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md#generating-2305-release-notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

<!--
To help with the large amounts of pull requests, we would appreciate your
reviews of other pull requests, especially simple package updates. Just leave a
comment describing what you have tested in the relevant package/service.
Reviewing helps to reduce the average time-to-merge for everyone.
Thanks a lot if you do!

List of open PRs: https://github.com/NixOS/nixpkgs/pulls
Reviewing guidelines: https://nixos.org/manual/nixpkgs/unstable/#chap-reviewing-contributions
-->

rogarb committed:
inotify-tools: add fanotifySupport option

This option allows to build additional utils, fsnotifywait and
fsnotifywatch, which use the fanotify kernel API.

The fanotify support is enabled by default since it depends on
CONFIG_FANOTIFY which is enabled in the default config of the default
NixOS kernel.

rogarb commented:
@SCOTT-HAMILTON @MarcWeber @pSub Did you have any chance to review the changes?

pSub reviewed (approved):
LGTM

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/prs-already-reviewed/2617/977

rogarb commented on pkgs/development/tools/misc/inotify-tools/default.nix:
```diff
@@ -11,6 +11,10 @@ stdenv.mkDerivation rec {
   };
 
+  configureFlags = [
+    (lib.enableFeature fanotifySupport "fanotify")
```
I don't see any reason why not to enable by default.
I didn't enable it by default because I didn't want to change the default behaviour.
As it depends on a kernel feature (CONFIG_FANOTIFY, which is available since kernel 5.9 according to [1](https://github.com/inotify-tools/inotify-tools/pull/134)), it might be better to condition it to this feature being activated in the kernel config. For the record, it is enabled in the NixOS kernel pkgs.linuxKernel.packages.linux_latest_libre default configuration which I am using.

rogarb replied:
I just checked and it is seems to be actually enabled in the default common kernel configuration: [https://github.com/NixOS/nixpkgs/blob/e6db435973160591fe7348876a5567c729495175/pkgs/os-specific/linux/kernel/common-config.nix#L319](https://github.com/NixOS/nixpkgs/blob/e6db435973160591fe7348876a5567c729495175/pkgs/os-specific/linux/kernel/common-config.nix#L319)

rogarb replied:
What do you think about enabling fanotifySupport by default and:
```suggestion
    (lib.enableFeature (fanotifySupport && linux.kernelAtLeast "5.9") "fanotify")
```
I couldn't find a way to check the specific "CONFIG_FANOTIFY" status: 
```
$ zgrep FANOTIFY /proc/config.gz
CONFIG_FANOTIFY=y
CONFIG_FANOTIFY_ACCESS_PERMISSIONS=y
```
but `linux.config.isEnabled "FANOTIFY"` yields a `false` value.

SuperSandro2000 replied:
> For the record, it is enabled in the NixOS kernel pkgs.linuxKernel.packages.linux_latest_libre default configuration which I am using.

This would be depended on the normal kernel since libre and hardened are provided on a best effort basis and where in the past just marked broken because they where badly maintained.



> I just checked and it is seems to be actually enabled in the default common kernel configuration:

Then from my side we can just activate this. We still have the 4_14 ad 5_4 kernel https://github.com/NixOS/nixpkgs/blob/master/pkgs/top-level/linux-kernels.nix though. Not sure if it is realistic that someone is using them.

> What do you think about enabling fanotifySupport by default and:

This would likely always return true since linux points to the default kernel. Also integrating nixos options into packages directly is not possible.

rogarb replied:
Ok, I just sent a new version with the option enabled by default.
I guess it is still better to present it as an option which can be deactivated using pkg.override function.

rogarb commented:
@SuperSandro2000 Are you fine with this new version?

rogarb committed:
inotify-tools: refactor using `finalAttrs` syntax

The `finalAttrs` syntax allows to get rid of using recursive attribute
sets for package definition.

https://nixos.org/manual/nixpkgs/unstable/#mkderivation-recursive-attributes

rogarb committed:
inotify-tools: refactor fetchFromGitHub attributes

The documentation currently recommends to use `hash` instead of
`sha256`.

https://nixos.org/manual/nixpkgs/unstable/#fetchfromgithub

SuperSandro2000 reviewed (commented):
one small change otherwise LGTM
